lib/utils.js

```diff
@@ -251,9 +251,43 @@ function getPackageVersion (pkgName) {
   return pkgInfo.version;
 }
 
+/**
+ * Pulls the initial values of Appium settings from the given capabilities argument.
+ * Each setting item must satisfy the following format:
+ * `setting[setting_name]: setting_value`
+ * The capabilities argument itself gets mutated, so it does not contain parsed
+ * settings anymore to avoid further parsing issues.
+ *
+ * @param {?Object} caps - Capabilities dictionary. It is mutated if
+ * one or more settings have been pulled from it
+ * @returns {Object} - An empty dictionary if the given caps contains no
+ * setting items or a dictionary containing parsed Appium setting names along with
+ * their values.
+ */
+function pullSettings (caps) {
+  if (!_.isPlainObject(caps) || _.isEmpty(caps)) {
+    return {};
+  }
+
+  const result = {};
+  for (const [key, value] of _.toPairs(caps)) {
+    const match = /\bsettings\[(\S+)\]$/.exec(key);
+    if (!match) {
+      if (_.isPlainObject(value)) {
+        Object.assign(result, pullSettings(value));
```

Comment:
i'm having trouble imagining the need for a recursive case...?

Reply:
this might be the case for w3c caps, where they have firstMatch and alwaysMatch properties

Reply:
I've changed the logic there. Now it's no more recursive